Don't emit permissions for expired/not-yet-valid conditions

HannesWell opened the pull request:
At the moment Passage's access-cycle does not check if a condition is expired or has not yet started. As a consequence Passage grants access to features although the corresponding grant is expired or has not yet started. The test `AccessAcquireTest.acquireLicenseGrant()` demonstrates this, since it passes although the grant for the `prince-to-frog` feature expired in December of last year.

Checking the validity period is IMHO a crucial feature of a library like passage.

The most suitable location to implement this seemed to be the `BasePermissionEmittingService.emitFor()` because the hardware evaluation instructions are checked there as well. Then that methods check the temporal and 'local' validity of a condition.
Please let me know if there is another more suitable place to implement the check.

eparovyshnaya reviewed (changes requested):
Dear @HannesWell 

code looks good. 
It fails tests, so either tests must be reasonably changed, or the code.

HannesWell commented:
> Dear @HannesWell
> 
> code looks good. It fails tests, so either tests must be reasonably changed, or the code.

Thank you @eparovyshnaya.

There was indeed a bug in the implementation. Basically the angular-bracket was wrong way round (of course it is only invalid if `now < from || expired < now`).
I also reworked the code a bit since _now_ cannot be outside of both borders of the validity period interval so a list does not make sense here and an Optional is more suitable and makes the code simpler.

Additionally I added test cases to ensure that the validity-period check works correctly.

HannesWell commented:
Hi @eparovyshnaya, could you please provide the updated encrypted license files or (if possible) advice me how to re-create them, so that we can proceed with this PR?

HannesWell commented:
> Hi @eparovyshnaya, could you please provide the updated encrypted license files or (if possible) advice me how to re-create them, so that we can proceed with this PR?

Hello @eparovyshnaya 

I just created new key-pairs for the test projects and encrypted the updated license files with them to make the tests pass without modification. The new key-pairs as well as the code to create the encrypted file is part of the PR.
Just let me know if you want me if you want to keep them in the test plug-ins.

Thank you.

HannesWell commented:
@eparovyshnaya or @ruspl-afed could you please be so kind to review this PR.

eparovyshnaya commented on tests/org.eclipse.passage.lbc.base.tests/src/org/eclipse/passage/lbc/base/tests/EncryptTestKeys.java:
```diff
@@ -0,0 +1,67 @@
+package org.eclipse.passage.lbc.base.tests;
```
Can you be so kind to explain the purpose of this?
This is neither _final_ nor _abstract_ class, full of static methods and, surprisingly, posessing _main_. For what I see, it's unused. If you meant some testing necessities to be satisfied here, that would be great to convey them into accordance with coding principles of the project.

HannesWell replied:
I used this code to create encrypt the modified `.lic` to get the new `.licen` file. That is the only purpose and the reason it has a main. It is not needed to execute the tests.
I thought you might be interested, in case some license files used for testing have to be changed and encrypted again.
But I'm also fine with removing it from this PR.

eparovyshnaya commented on bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/conditions/evaluation/BasePermissionEmittingService.java:
```diff
@@ -149,15 +172,15 @@ private ServiceInvocationResult<Emission> fail(ConditionPack pack, TroubleCode c
 
 	private ZonedDateTime from(ValidityPeriod period) {
-		if (ValidityPeriodClosed.class.isInstance(period)) {
+		if (period instanceof ValidityPeriodClosed) {
```
What is the difference that makes you do the alternation?

HannesWell replied:
From my experience instanceof is the usual way to check for a type in cases like this and for example Sonar-Lint says the same.
But if you prefer to keep it as it is, I can revert it.

eparovyshnaya replied:
It's better for MR to stay focused on the task and not to contain irrelevant changes. Please revert.

eparovyshnaya commented on tests/org.eclipse.passage.lic.base.tests/src/org/eclipse/passage/lic/internal/base/tests/conditions/evaluation/BasePermissionEmittingServiceTest.java:
```diff
@@ -171,7 +189,7 @@ private Condition corrupted() {
 	}
 
-	private Condition humanoid(int legs, int eyes) {
+	private FakeCondition humanoid(int legs, int eyes) {
```
Is there a reason to narrowing down return type here?

HannesWell replied:
Yes, otherwise the method `withValidityPeriod()` cannot be called on the returned object in the two new test cases.

eparovyshnaya commented on tests/org.eclipse.passage.seal.demo.tests/OSGI-INF/anti-human-magic.product_0.2.1.keys_xmi:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="ASCII"?>
```
This is a testing resource and must reside in _resource_ folder of curresponding testing bundle

HannesWell replied:
Done.

eparovyshnaya reviewed (changes requested):
Dear @HannesWell thank you for your work. 
Please pay attention to the comments below.

eparovyshnaya commented on bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/conditions/evaluation/BasePermissionEmittingService.java:
```diff
@@ -113,9 +117,28 @@ private ServiceInvocationResult<Emission> emitFor(Condition condition, Condition
+		} else if (now.isAfter(end)) {
+			troubles = Optional.of(new LicenseExpired());
+		}
+		return troubles.map(c -> new Trouble(c, c.explanation()));
```
Code explanation is static regarding the situation and thus cannot be used to describe an actual trouble at hand. Use dynamic info with actual details that will help a user to define reason of denial.

eparovyshnaya replied:
return empty Optional if no troubles found

eparovyshnaya commented on bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/conditions/evaluation/BasePermissionEmittingService.java:
```diff
@@ -113,9 +117,28 @@ private ServiceInvocationResult<Emission> emitFor(Condition condition, Condition
+		ZonedDateTime now = ZonedDateTime.now();
+		ZonedDateTime start = from(condition.validityPeriod()); // TODO: #566015
+		ZonedDateTime end = expiration(condition.validityPeriod());
+		Optional<TroubleCode> troubles = Optional.empty();
```
excessive var, vails the logic.

eparovyshnaya commented on bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/conditions/evaluation/BasePermissionEmittingService.java:
```diff
@@ -113,9 +117,28 @@ private ServiceInvocationResult<Emission> emitFor(Condition condition, Condition
+		ZonedDateTime end = expiration(condition.validityPeriod());
+		Optional<TroubleCode> troubles = Optional.empty();
+		if (now.isBefore(start)) {
+			troubles = Optional.of(new LicenseNotStarted());
```
return here

eparovyshnaya commented on bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/conditions/evaluation/BasePermissionEmittingService.java:
```diff
@@ -113,9 +117,28 @@ private ServiceInvocationResult<Emission> emitFor(Condition condition, Condition
+		if (now.isBefore(start)) {
+			troubles = Optional.of(new LicenseNotStarted());
+		} else if (now.isAfter(end)) {
+			troubles = Optional.of(new LicenseExpired());
```
return here as well

eparovyshnaya commented on tests/org.eclipse.passage.seal.demo.tests/resource/anti-human-magic.product_0.2.1.keys_xmi:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="ASCII"?>
```
Look at how licenses are stored. There should probably be a `resource/keys` folder for keys

eparovyshnaya commented:
The issue #1136 is fixed with MRs #1138 and #1139
